// @nestjs-plus/rabbitmq/lib/rabbitmq.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
var RabbitMQModule_1;
"use strict";
const discovery_1 = require("@nestjs-plus/discovery");
const common_1 = require("@nestjs/common");
const common_2 = require("@nestjs-plus/common");
const external_context_creator_1 = require("@nestjs/core/helpers/external-context-creator");
const lodash_1 = require("lodash");
const connection_1 = require("./amqp/connection");
const rabbitmq_constants_1 = require("./rabbitmq.constants");
let RabbitMQModule = RabbitMQModule_1 = class RabbitMQModule {
    constructor(discover, amqpConnection, externalContextCreator) {
        this.discover = discover;
        this.amqpConnection = amqpConnection;
        this.externalContextCreator = externalContextCreator;
        this.logger = new common_1.Logger(RabbitMQModule_1.name);
    }
    static forRootAsync(asyncOptionsFactoryProvider) {
        return {
            module: RabbitMQModule_1,
            exports: [connection_1.AmqpConnection],
            imports: asyncOptionsFactoryProvider.imports,
            providers: [
                ...this.createAsyncProviders(asyncOptionsFactoryProvider),
                {
                    provide: connection_1.AmqpConnection,
                    useFactory: async (config) => {
                        const connection = new connection_1.AmqpConnection(config);
                        await connection.init();
                        const logger = new common_1.Logger(RabbitMQModule_1.name);
                        logger.log('Successfully connected to RabbitMQ');
                        return connection;
                    },
                    inject: [rabbitmq_constants_1.RABBIT_CONFIG]
                }
            ]
        };
    }
    static forRoot(config) {
        return {
            module: RabbitMQModule_1,
            providers: [
                {
                    provide: connection_1.AmqpConnection,
                    useFactory: async () => {
                        const connection = new connection_1.AmqpConnection(config);
                        await connection.init();
                        const logger = new common_1.Logger(RabbitMQModule_1.name);
                        logger.log('Successfully connected to RabbitMQ');
                        return connection;
                    }
                }
            ],
            exports: [connection_1.AmqpConnection]
        };
    }
    static build(config) {
        const logger = new common_1.Logger(RabbitMQModule_1.name);
        logger.warn('build() is deprecated. use forRoot() or forRootAsync() to configure RabbitMQ');
        return {
            module: RabbitMQModule_1,
            providers: [
                {
                    provide: connection_1.AmqpConnection,
                    useFactory: async () => {
                        const connection = new connection_1.AmqpConnection(config);
                        await connection.init();
                        const logger = new common_1.Logger(RabbitMQModule_1.name);
                        logger.log('Successfully connected to RabbitMQ');
                        return connection;
                    }
                }
            ],
            exports: [connection_1.AmqpConnection]
        };
    }
    static attach(connection) {
        return {
            module: RabbitMQModule_1,
            providers: [
                {
                    provide: connection_1.AmqpConnection,
                    useValue: connection
                }
            ],
            exports: [connection_1.AmqpConnection]
        };
    }
    async onModuleInit() {
        this.logger.log('Initializing RabbitMQ Handlers');
        const rabbitMeta = await this.discover.providerMethodsWithMetaAtKey(rabbitmq_constants_1.RABBIT_HANDLER);
        const grouped = lodash_1.groupBy(rabbitMeta, x => x.discoveredMethod.parentClass.name);
        const providerKeys = Object.keys(grouped);
        for (const key of providerKeys) {
            this.logger.log(`Registering rabbitmq handlers from ${key}`);
            await Promise.all(grouped[key].map(async ({ discoveredMethod, meta: config }) => {
                const handler = this.externalContextCreator.create(discoveredMethod.parentClass.instance, discoveredMethod.handler, discoveredMethod.methodName);
                const { exchange, routingKey, queue } = config;
                this.logger.log(`${discoveredMethod.parentClass.name}.${discoveredMethod.methodName} {${config.type}} -> ${exchange}::${routingKey}::${queue ||
                    'amqpgen'}`);
                return config.type === 'rpc'
                    ? this.amqpConnection.createRpc(handler, config)
                    : this.amqpConnection.createSubscriber(handler, config);
            }));
        }
    }
    static createAsyncProviders(asyncOptionsFactoryProvider) {
        const optionsProvider = common_2.createAsyncOptionsProvider(rabbitmq_constants_1.RABBIT_CONFIG, asyncOptionsFactoryProvider);
        if (asyncOptionsFactoryProvider.useFactory) {
            return [optionsProvider];
        }
        if (asyncOptionsFactoryProvider.useClass) {
            return [
                optionsProvider,
                {
                    provide: asyncOptionsFactoryProvider.useClass,
                    useClass: asyncOptionsFactoryProvider.useClass
                }
            ];
        }
        if (asyncOptionsFactoryProvider.useExisting) {
            return [
                optionsProvider,
                {
                    provide: asyncOptionsFactoryProvider.useExisting.provide ||
                        asyncOptionsFactoryProvider.useExisting.value.constructor.name,
                    useValue: asyncOptionsFactoryProvider.useExisting.value
                }
            ];
        }
        return [];
    }
};
RabbitMQModule = RabbitMQModule_1 = __decorate([
    common_1.Module({
        imports: [discovery_1.DiscoveryModule]
    }),
    __metadata("design:paramtypes", [discovery_1.DiscoveryService,
        connection_1.AmqpConnection,
        external_context_creator_1.ExternalContextCreator])
], RabbitMQModule);
exports.RabbitMQModule = RabbitMQModule;
